src/devices/src/bus.rs

```diff
@@ -197,21 +197,32 @@ mod tests {
     fn bus_insert() {
         let mut bus = Bus::new();
         let dummy = Arc::new(Mutex::new(DummyDevice));
-        assert!(bus.insert(dummy.clone(), 0x10, 0).is_err());
-        assert!(bus.insert(dummy.clone(), 0x10, 0x10).is_ok());
-
-        let result = bus.insert(dummy.clone(), 0x0f, 0x10);
-        assert!(result.is_err());
-        assert_eq!(format!("{:?}", result), "Err(Overlap)");
-
-        assert!(bus.insert(dummy.clone(), 0x10, 0x10).is_err());
-        assert!(bus.insert(dummy.clone(), 0x10, 0x15).is_err());
-        assert!(bus.insert(dummy.clone(), 0x12, 0x15).is_err());
-        assert!(bus.insert(dummy.clone(), 0x12, 0x01).is_err());
-        assert!(bus.insert(dummy.clone(), 0x0, 0x20).is_err());
-        assert!(bus.insert(dummy.clone(), 0x20, 0x05).is_ok());
-        assert!(bus.insert(dummy.clone(), 0x25, 0x05).is_ok());
-        assert!(bus.insert(dummy.clone(), 0x0, 0x10).is_ok());
+        for tc in &vec![
+            // insert len should not be 0.
+            (0x10, 0, false, "Err(Overlap)"),
+            (0x10, 0x10, true, ""),
+            // this overlaps to address space of exist bus device based on 0x10
+            (0x0f, 0x10, false, "Err(Overlap)"),
+            // this overlaps to address space of exist bus device based on 0x10
+            (0x10, 0x10, false, "Err(Overlap)"),
+            // this overlaps to address space of exist bus device based on 0x10
+            (0x10, 0x15, false, "Err(Overlap)"),
```

Comment:
I would be in favor of removing some of these test cases, as they overlap in behavior.

For example, after adding a device at address 0x10 with len 0x10, it would be enough to test that an overlap occurs with 2 test cases:
 - a device with start address in the interval (0x10, 0x20)
 - a device with start address in (0x0, 0x9) and a len that would make it overlap the (0x10, 0x20) interval.

Similar reasoning for cases should be ok.

Reply:
agree, I also feel they looks a little superfluous. How about below?
```rust
vec![
            // insert len should not be 0.
            (0x10, 0, false, "Err(Overlap)"),
            (0x10, 0x10, true, ""),
            // This overlaps the address space of the existing bus device at 0x10.
            (0x08, 0x10, false, "Err(Overlap)"),
            // This overlaps the address space of the existing bus device at 0x10.
            (0x10, 0x20, false, "Err(Overlap)"),
            (0x0, 0x10, true, ""),
            (0x20, 0x05, true, ""),
        ]
```

Reply:
Looks good!